### ISSUE TITLE
Naming convention for JavaScript events

### ISSUE THREAD

**Issue by Brockenstein** [tmpworldwide/tmpworldwide.github.io#33]
Posting this in behalf of @dchacon1 

I would like to suggest a naming convention for classes that have an event bound to them.  In addition to a naming convention, I also suggest any class that has an event bound to it not have any CSS styling on it. Doing this will help to easily recognize and separate JavaScript from CSS. There are many advantages to this. 
 
Here is an example of not separating JavaScript and CSS:  
 
`<a class="button fancy-button" href=”#example”>Read More</a>` 

In this example you are not able to easily tell what class the JS event is bound to – or if it is even bound to a class. The JS target could be the button class or apply-button class or even the anchor tag itself. 
In addition to this, you also don’t know what exactly will happen if you click this anchor tag – you don’t know if it will expand a hidden section, if it will scroll you to a different part of the page, or do something else. 
 
Here is a suggested solution: 
 
```
<a class="button fancy-button js-expand" href=”#example”>Read more</a>
 
<a class="button fancy-button js-scroll" href=”#example”>Read more</a>
 
<a class="button fancy-button" href=”/search-jobs”>Read more</a>
```
 
By adding a “js” in front of a class name you can clearly define what class the JS event is bound to. The advantage of this is that you can use the same anchor tag and classes on multiple pages without having to worry about console errors. If you need the anchor tag to look exactly the same but have different functionality you can simply remove the JavaScript class and change it to another one as – you can see this done in the second example.
 
If you do not want any JavaScript functionality on the anchor tag but still want it to look identical, simply remove it and add a link in the href attribute.
 
 
 
In conclusion:
It is my suggestion that this practice become a standard guideline for developers not only at TMP but everywhere. When you style a class that has an event bound to it the class becomes much more difficult to reuse. In many cases the styles of the class will be duplicated under a different class name so that they can be used on a different page. 
 
If we put into practice this new standard it will help reduce duplicating CSS that you can’t reuse because it has an event bound to it . It will also help clearly define in the HTML what elements are being targeted by JS – which also reduces maintenance time for future developers who may have to diagnose potential JS errors.

**Comment by mhartz** [comment]
Heads up - I still see some updates :P.

Either way, why target classes? You can target data attributes or even create your own.

```
<a href="#item" data-expanded="true">Expand Me!</a>
<a href="#item" expanded-true>Expand me!</a>

a[data-expanded="true"] { display: block; }
a[expanded-true] { display: block; }
```

The good thing is that you can use boolean to set that, along with other attributes at the same time such as aria labels.

**Comment by Brockenstein** [comment]
@mhartz  you will always be apart of my team :) 

I like what you are saying, @fwitkowski  recently brought that idea to my attention as well. @fwitkowski I would love to hear further insight from you into your experience on targeting for js.

**Comment by dchacon1** [comment]
@mhartz @Brockenstein 
Well-written naming conventions for classes are supposed to help developers identify the purpose of an element or give insight to an elements state.

Why use classes for JS? One reason put simply is for the same reason as above - we can distinguish what classes are being used for css vs what classes are being used for JS.

You may not have known this but TB already uses this principle.  
https://www.screencast.com/t/XbhsuTxtI083

This screencast came from https://jobs.utc.com/ 
Notice there is a class called "js" added to the html element. This is added by TB core. When a user disables JS the class is changed to "no-js". 

All I am saying is that we carry this principle to other areas of our coding.

An additional reason to use JS classes is so that styles are not being added to elements that have a JS listener - in effect separating JS and CSS in a more obvious way. Why does it matter?
Two reasons:
1) Diagnosis of JS bugs becomes easier because we will be able to know - just by looking at the html - if a bug is related to js or not.
2) If you are using styles on a class that also has a JS listener then that class is no longer as reusable as it could be. Separating the class so that there is a js class and a css class will make it more reusable.

**Comment by Brockenstein** [comment]
I think that is a valid point about making it clearer on what element has JS acting on it.  This is especially important to know as we have seen default classes changed or removed that had JS acting on them but the developer didn't know. I find that oftentimes data attributes are hidden in TBs razor. So it is hard to identify what the css is targeting but I would love to hear more from @fwitkowski
